cmd/firectl/main.go

```diff
@@ -181,26 +197,85 @@ func main() {
 		Debug:             opts.Debug,
 	}
 
-	m, err := firecracker.NewMachine(fcCfg, firecracker.WithLogger(log.NewEntry(logger)))
+	ctx := context.Background()
+	cmd, cleanup, err := buildCommand(ctx, fcCfg.SocketPath, opts)
+	if err != nil {
+		log.Fatalf("Failed to build command: %v", err)
+	}
+	defer cleanup()
+
+	m, err := firecracker.NewMachine(cmd, fcCfg, firecracker.WithLogger(log.NewEntry(logger)))
 	if err != nil {
 		log.Fatalf("Failed creating machine: %s", err)
 	}
+	defer m.Close()
 
-	ctx := context.Background()
 	vmmCtx, vmmCancel := context.WithCancel(ctx)
 	defer vmmCancel()
-	errchan, err := m.Init(ctx)
+
+	errCh, err := m.Init(ctx)
 	if err != nil {
 		log.Fatalf("Firecracker Init returned error %s", err)
-	} else {
-		m.StartInstance(vmmCtx)
 	}
+	m.StartInstance(vmmCtx)
 
 	// wait for the VMM to exit
-	err = <-errchan
-	if err != nil {
+	if err := <-errCh; err != nil {
 		log.Fatalf("startVMM returned error %s", err)
-	} else {
-		log.Printf("startVMM was happy")
 	}
+	log.Printf("startVMM was happy")
+}
+
+func buildCommand(ctx context.Context, socketPath string, opts options) (firecracker.Command, func() error, error) {
+	builder := firecracker.VMCommandBuilder{}.
+		WithBin(opts.FcBinary).
+		WithSocketPath(socketPath)
+
+	fn := func() error {
+		return nil
+	}
+
+	switch opts.FcConsole {
+	case ConsoleXterm:
+		bin := builder.Bin()
+		args := builder.Args()
+		builder = builder.WithBin(terminalProgram).
+			WithArgs(append([]string{"-e", bin}, args...))
```

Comment:
I'm not sure if this approach is better, as for me, extra level of indirection makes things harder to understand.
Consider this version:
```go
	switch m.cfg.Console {
	case ConsoleXterm:
		cmd = exec.CommandContext(ctx, xtermPath, "-e", m.binPath)
	case ConsoleStdio:
		cmd = exec.CommandContext(ctx, m.binPath)
		cmd.Stdin = os.Stdin
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
	case ConsoleFile:
		stdout, stdout, err := openLogFiles(...)
		if err != nil {
			...
		}
		cmd = exec.CommandContext(ctx, m.binPath)
		cmd.Stdout = stdout
		cmd.Stderr = stderr
	default:
		cmd = exec.CommandContext(ctx, m.binPath)
	}
	
	cmd.Args = append(cmd.Args, "--api-sock", m.cfg	.SocketPath)
```

Reply:
I only changed the `ConsoleXterm` case as that one was a little funky. I've kept the other ones the same aren't difficult to read.